hls4ml/model/optimizer/passes/hgq_proxy_model.py

```diff
@@ -0,0 +1,151 @@
+import re
+from warnings import warn
+
+from hls4ml.backends.fpga.fpga_types import NamedType
+from hls4ml.model.layers import Layer, register_layer
+from hls4ml.model.optimizer import OptimizerPass, register_pass
+from hls4ml.model.types import FixedPrecisionType, UnspecifiedPrecisionType, WeightVariable
+
+re_purge_prefix = re.compile(r'(?<!\w)(?:ap_|ac_)', re.IGNORECASE)
+re_parse_fixed = re.compile(r'\s*(u?)fixed<([^>]+)>\s*', re.IGNORECASE)
+
+
+class FixedPointQuantizer(Layer):
+    def initialize(self):
+        inp = self.get_input_variable()
+        shape = inp.shape
+        dims = inp.dim_names
+        self.add_output_variable(shape, dims)
+        self.set_attr('n_in', self.get_input_variable().size())
+        self.overrides = self.attributes['overrides']
+        self.fusible = self.attributes['fusible']
+        self.SAT, self.RND = self.attributes['SAT'], self.attributes['RND']
+        self.mask_kbi = self.attributes.get('mask_kbi', None)
+
+
+class UnaryLUT(Layer):
+    def initialize(self):
+        inp = self.get_input_variable()
+        shape = inp.shape
+        dims = inp.dim_names
+        self.add_output_variable(shape, dims)
+        self.set_attr('n_in', inp.size())
+        self.table = self.attributes['table']
+        self.table_size = self.attributes['table_size']
+
+        table_t = to_hls4ml_fixed(self.attributes['table_t'])
+        self.add_weights_variable(name='table', var_name='table{index}', precision=table_t, data=self.table)
+
+
+def to_hls4ml_fixed(fixed: str):
+    matched = re_parse_fixed.match(re_purge_prefix.sub('', fixed))
+    assert matched is not None, f'Cannot parse {fixed}'
+    signed = matched.group(1) != 'u'
+    b, i, *args = matched.group(2).split(',')
+    b, i = int(b), int(i)
+    args = [arg.upper() for arg in args]
+    new_type = FixedPrecisionType(b, i, signed, *args)
+    # For some reason, __class__ is overwritten in hls4ml
```

Comment:
I wonder if that's the conversion to a backend version, so that it knows how to print it properly. (I am just guessing now.)

Reply:
Unfortunately, this need to go through another round of meta programming (done in `TransformTypes` optimizer) to got printed well, thus unlikely can be used at the spot of code generation (final string produced at the spot).

Reply:
Ok, glad to see that the backend specific types are no longer generated on the fly. Though, the current serialization strategy seems to be a bit problematic: Say one defines `SAT` without rounding strategy, it will generate `ap_(u)fixed<w,i,AP_SAT>`, which doesn't do what it intends to do.... can be fixed quite easily though.

To do it properly, probably, we will the backends declaring their required type string generator (converter, whatever). As I went to through the backends, I did not spot them at the moment, and it seems that everything is still handled through an optimizer pass. This will go to a separate PR as it would require changing working flows irrelevant to this.